### PR TITLE
Add Django users to groups

### PR DESCRIPTION
**Pull request by m-bull**
Add functionality to add users to groups if they are specified in `pulp_django_users`.

### CODE REVIEW

File: roles/pulp_django_user/tasks/user_groups.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+- name: Add user to group
+  uri:
+    url: "{{ pulp_url }}{{ result.json.results[0].pulp_href }}users/"
+    user: "{{ pulp_server_admin_user }}"
+    password: "{{ pulp_server_admin_password }}"
+    body:
+      username: "{{ item.username }}"
+    body_format: json
+    method: POST
+    status_code: 201
```
Comment (markgoddard) [inline roles/pulp_django_user/tasks/user_groups.yaml:19]: Is it idempotent? Will it be 201 only if created?

File: roles/pulp_django_user/tasks/user_groups.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  uri:
+    url: "{{ pulp_add_group_user_url }}?name={{ item.group }}"
+    user: "{{ pulp_admin_username }}"
+    password: "{{ pulp_admin_password }}"
+    method: GET
+    status_code: 200
+
+- name: Add user to group
+  uri:
+    url: "{{ pulp_url }}{{ result.json.results[0].pulp_href }}users/"
```
Comment (markgoddard) [inline roles/pulp_django_user/tasks/user_groups.yaml:12]: result undefined?

File: roles/pulp_django_user/tasks/user_groups.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+---
```
Comment (markgoddard) [inline roles/pulp_django_user/tasks/user_groups.yaml:1]: Shame there's no squeezer module, but probably not worth the bother of writing one for this.

File: roles/pulp_django_user/tasks/main.yml [hunk 4]
```diff
@@ -46,3 +46,14 @@
...
+- name: Add user to group(s)
+  include_tasks: user_groups.yaml
+  loop: >-
+       {%- set user_groups = [] -%}
+       {%- for user in pulp_django_users -%}
+       {%- for group in user.groups -%}
+       {{ user_groups.append({'username': user.username, 'group': group }) }}
+       {%- endfor -%}
+       {%- endfor -%}
+       {{ user_groups }}
```
Comment (markgoddard) [inline roles/pulp_django_user/tasks/main.yml:59]: nit: more idiomatic ansible might use

```
with_subelements:
  - "{{ pulp_django_users }}"
  - groups
```

This is fine though

File: roles/pulp_django_user/tasks/main.yml [hunk 5]
```diff
@@ -46,3 +46,14 @@
   loop: "{{ pulp_django_users }}"
   loop_control:
     label: "{{ item.username }}"
+
+- name: Add user to group(s)
+  include_tasks: user_groups.yaml
+  loop: >-
+       {%- set user_groups = [] -%}
+       {%- for user in pulp_django_users -%}
+       {%- for group in user.groups -%}
```
Comment (markgoddard) [inline roles/pulp_django_user/tasks/main.yml:55]: Allow groups to be undefined?

**Review by markgoddard — CHANGES_REQUESTED** [review]

File: roles/pulp_django_user/tasks/main.yml (hunk 4 above)
Comment (m-bull) [reply roles/pulp_django_user/tasks/main.yml:59]: New to me - thanks!

File: roles/pulp_django_user/tasks/user_groups.yaml (hunk 1 above)
Comment (m-bull) [reply roles/pulp_django_user/tasks/user_groups.yaml]: Always 201, better logic in place for this now.

File: roles/pulp_django_user/tasks/user_groups.yaml (hunk 2 above)
Comment (m-bull) [reply roles/pulp_django_user/tasks/user_groups.yaml]: Urgh, fixed this.

File: roles/pulp_django_user/tasks/main.yml (hunk 4 above)
Comment (m-bull) [reply roles/pulp_django_user/tasks/main.yml]: Sadly did away with all of this tremendously horrible jinja.

File: roles/pulp_django_user/tasks/main.yml (hunk 5 above)
Comment (m-bull) [reply roles/pulp_django_user/tasks/main.yml]: Fixed now.

File: roles/pulp_django_user/tasks/user_groups/add_or_remove_users.yml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    force_basic_auth: true
+  register: user_group_result
+
+- name: Set current_group_membership fact
+  set_fact:
+    current_group_membership: "{{ user_group_result | json_query('json.results[*].groups') | flatten | map(attribute='name') | list }}"
+
+- name: Remove user from groups not defined in pulp_django_users
+  include_tasks: remove_user_from_groups.yml
+  loop: "{{ current_group_membership | difference(user.groups | default([])) }}"
```
Comment (markgoddard) [inline roles/pulp_django_user/tasks/user_groups/add_or_remove_users.yml:19]: nit: I think you need `default([], true)` to make this work:

> Adding an empty `groups:` for a user will result in that user being removed from all groups.

File: roles/pulp_django_user/tasks/user_groups/add_or_remove_users.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  loop_control:
+    loop_var: remove_group
+
+- name: Add user to groups defined in pulp_django_users
+  include_tasks: add_user_to_groups.yml
+  loop: "{{ user.groups | default([]) | difference(current_group_membership) }}"
```
Comment (markgoddard) [inline roles/pulp_django_user/tasks/user_groups/add_or_remove_users.yml:25]: ditto

**Review by markgoddard — CHANGES_REQUESTED** [review]

File: roles/pulp_django_user/tasks/user_groups/add_or_remove_users.yml (hunk 6 above)
Comment (m-bull) [reply roles/pulp_django_user/tasks/user_groups/add_or_remove_users.yml]: You're right - fixed in https://github.com/stackhpc/ansible-collection-pulp/pull/31/commits/2794ce64d33d9c63fddc38636d9a89dd0cfba397

File: roles/pulp_django_user/tasks/user_groups/add_or_remove_users.yml (hunk 7 above)
Comment (m-bull) [reply roles/pulp_django_user/tasks/user_groups/add_or_remove_users.yml]: ditto https://github.com/stackhpc/ansible-collection-pulp/pull/31/commits/2794ce64d33d9c63fddc38636d9a89dd0cfba397

**Review by markgoddard — APPROVED** [review]

**Pull request merged**